Biot-Savart volume integral

ddudt opened the pull request:
Computes B_plasma in a volume as: 

𝐁ᵥ(𝐫) = μ₀/4π ∫ 𝐉(𝐫') × (𝐫 − 𝐫')/|𝐫 − 𝐫'|³ d³𝐫'

This is useful for when you need B_total = B_plasma + B_external in the full computational domain, not just on the boundary surface.

daniel-dudt committed:
initial commit

ddudt committed:
Merge branch 'master' into dd/B_plasma

daniel-dudt committed:
do calculations in Cartesian coordinates

dpanici commented:
use this fxn https://github.com/PlasmaControl/DESC/blob/b10f8c8be9d3e751684e5d34abecaf5cd40b2205/desc/magnetic_fields/_core.py#L31

dpanici committed:
Merge branch 'master' into dd/B_plasma

ddudt committed:
Merge branch 'master' into dd/B_plasma

dpanici commented:
`scipy.integrate.tpl_quad` to check this

dpanici commented:
[IAEA2020_JGeiger_Manuscript_8p_finalversion (1).pdf](https://github.com/user-attachments/files/16090258/IAEA2020_JGeiger_Manuscript_8p_finalversion.1.pdf)
[Contributions to Plasma Physics - 2010 - Geiger - Effects of Net Currents on the Magnetic Configuration of W7‐X (2).pdf](https://github.com/user-attachments/files/16090260/Contributions.to.Plasma.Physics.-.2010.-.Geiger.-.Effects.of.Net.Currents.on.the.Magnetic.Configuration.of.W7.X.2.pdf)

f0uriest commented:
Several people at the scidac meeting today are using the BMW code: https://github.com/ORNL-Fusion/BMW/tree/master

refs: https://scipub.euro-fusion.org/wp-content/uploads/eurofusion/WPS1CPR17_17542_submitted-4.pdf
https://bpb-us-e2.wpmucdn.com/wordpress.auburn.edu/dist/5/118/files/2022/05/ISHW2017.pdf

basically it does a biot-savart volume integral to compute A (not B), then takes the finite difference curl of A to get B. As far as I can tell from the code the source grid is hardcoded to ntheta=101, ns, nzeta taken from the vmec/mgrid file. Allegedly it works well and gives accurate results?

Could try something similar (and maybe use spectral derivatives for going A->B)

ddudt committed:
Merge branch 'master' into dd/B_plasma

ddudt commented:
Try implementing Eq. 17-18 of [Peraza-Rodriguez et al. _Phys. Plasmas_ 2017](https://doi-org.ezproxy.princeton.edu/10.1063/1.4986447)
It should be a simple linear system solve to get the Fourier-Zernike coefficients of the vector potential as: 
$\nabla \times (\nabla \times \mathbf{A}) = \mu_0 \mathbf{J}$
And then the magnetic field is: 
$\mathbf{B} = \nabla \times \mathbf{A}$
And all of this can be done in the native $(\rho,\theta,\zeta)$ coordinate system.

daniel-dudt committed:
merge with master

daniel-dudt committed:
fix x_Equilibrium computation

daniel-dudt committed:
remove duplicate compute 'x' fun

ddudt committed:
Merge branch 'master' into dd/B_plasma

ddudt committed:
Merge branch 'master' into dd/B_plasma

ddudt committed:
Merge branch 'master' into dd/B_plasma

daniel-dudt committed:
fix norm axis bug

daniel-dudt committed:
Merge branch 'dd/B_plasma' of https://github.com/PlasmaControl/DESC into dd/B_plasma

daniel-dudt committed:
replace for loop with broadcasting

ddudt commented:
Update: I fixed a silly bug and now this is working. Direct Biot-Savart volume integral to compute B and the similar volume integral to compute A both agree well with BMW results at moderate resolution. But more benchmarking is needed.

rahulgaur104 commented:
Thank you! Could be really useful for my umbilic stuff!

ddudt committed:
Merge branch 'master' into dd/B_plasma

dpanici commented:
- show plots of error
- errors may be due to large matrix dot product errors?

daniel-dudt committed:
change np to jnp

ddudt committed:
Merge branch 'master' into dd/B_plasma

daniel-dudt committed:
change safenorm to linalg.norm

daniel-dudt committed:
Equilibrium subclass from _MagneticField

daniel-dudt committed:
remove IOAble parent from Equilibrium

ddudt committed:
Merge branch 'master' into dd/B_plasma

daniel-dudt committed:
revert new duplicate functions

daniel-dudt committed:
remove old functions

dpanici commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1205,6 +1211,126 @@ def need_src(name):
+        transforms=None,
+        chunk_size=None,
+    ):
+        """Compute magnetic field at a set of points.
```
be more specific: this is the magnetic field generated by the plasma currents.

I think also we could add a flag for external evaluation or not, that will allow the use of the virtual casing principle if desired (if one always knows the evaluation points are far from the plasma surface, like is the case with a lot of diagnostic flux loops)

dpanici commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1205,6 +1211,126 @@ def need_src(name):
+        eval_xyz = rpz2xyz(coords) if basis.lower() == "rpz" else coords
+        if source_grid is None:
+            source_grid = QuadratureGrid(
+                L=self.L_grid, M=self.M_grid, N=self.N_grid * self.NFP
```
same comment about nfp_loop

dpanici reviewed (changes requested):
- use nfp_loop
- I think worth adding an option for virtual casing principle

dpanici commented:
> Possibly redundant with #1436, should merge them once we decide on a public API.

I dont think it is redundant as #1436 does not expose this API at all to the user, it is done opaquely inside of the objective compute function. So there is no dependency there

dpanici committed:
Merge branch 'master' into dd/B_plasma

daniel-dudt committed:
replace nfp=1 grid with nfp_loop

ddudt committed:
Merge branch 'master' into dd/B_plasma

ddudt committed:
Merge branch 'master' into dd/B_plasma

maya-avida committed:
Plasma current magnetic field calculations outside plasma boundary (#1847)

This branch expands on dd/B_plasma to implement new two methods for
calculating the magnetic field outside the plasma boundary, including
taking the curl of the vector potential and the virtual casing method

---------

Co-authored-by: Daniel Dudt <33005725+ddudt@users.noreply.github.com>
Co-authored-by: daniel-dudt <daniel.dudt@princetonstellarators.energy>
Co-authored-by: Greta Hibbard <112410471+gretahibbard@users.noreply.github.com>
Co-authored-by: Greta Hibbard <gh7283@della-gpu.princeton.edu>
Co-authored-by: Rahul Gaur <19224702+rahulgaur104@users.noreply.github.com>
Co-authored-by: Rahul <rgaur@terpmail.umd.edu>
Co-authored-by: Greta Hibbard <gh7283@della9.princeton.edu>
Co-authored-by: Rory Conlin <wconlin@princeton.edu>
Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>
Co-authored-by: Yigit Gunsur Elmacioglu <102380275+YigitElma@users.noreply.github.com>
Co-authored-by: Kaya Unalmis <kayaunalmis@proton.me>

maya-avida committed:
Merge remote-tracking branch 'public_origin/master' into dd/B_plasma

maya-avida committed:
fixed import error in test_eq

maya-avida committed:
removed method flag from field_line_integrate

maya-avida committed:
Fixed grid test, also removed endpoint option from CylindricalGrid

maya-avida committed:
Fixed NotImplementedError test (since Chebyshev derivatives have now been implemented

maya-avida committed:
Merge branch 'master' into dd/B_plasma

maya-avida committed:
Prevent transform from warning about RPZ grids

maya-avida committed:
Merge branch 'dd/B_plasma' of https://github.com/PlasmaControl/DESC into dd/B_plasma

maya-avida committed:
changed current potential test for external_field error, since eq is now a magnetic field and therefore no longer raises the error

maya-avida committed:
Fixed unit test to give correct error

maya-avida committed:
Tests for eq.compute_magnetic_vector_potential and PlasmaField objets

maya-avida committed:
Somehow this image got swapped?

maya-avida committed:
fixing test docstring:

maya-avida committed:
Removed old r_endpoint and z_endpoint kwargs

maya-avida committed:
Removed old r_endpoint and z_endpoint kwargs

maya-avida committed:
Better tests for equilibrium vector potential

maya-avida committed:
More tests for PlasmaField and warnings for new transform methods.Also, CylindricalGrids can have 0 resolution now.

maya-avida committed:
Some more tests, and adding the new grid and basis classes to __all__

maya-avida committed:
added DoubleChebyshevFourierBasis to all

maya-avida committed:
Merge branch 'master' into dd/B_plasma

dpanici reviewed (changes requested):
partially reviewed but already some changes noted

ddudt committed:
Merge branch 'master' into dd/B_plasma

ddudt committed:
Merge branch 'master' into dd/B_plasma

YigitElma committed:
Merge branch 'master' into dd/B_plasma

ddudt committed:
Merge branch 'master' into dd/B_plasma

daniel-dudt committed:
fix merge conflicts

dario-panici commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1257,6 +1269,217 @@ def need_src(name):
+            B = integrate_surface(
+                eval_rpz, source_data, source_grid, kernel, chunk_size=chunk_size
+            )
+            if basis.lower == "xyz":
```
```suggestion
            if basis.lower() == "xyz":
```

dario-panici commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1257,6 +1269,217 @@ def need_src(name):
+                eval_rpz, source_data, source_grid, kernel, chunk_size=chunk_size
+            )
+            if basis.lower == "xyz":
+                B = rpz2xyz_vec(B, phi=coords[:, 1])
```
```suggestion
                B = rpz2xyz_vec(B, phi=eval_rpz[:, 1])
```
coords is xyz if basis is xyz
